### PR TITLE
raftstore: always enable entry cache for async write io

### PR DESCRIPTION
**Pull request by gengliqi**
Signed-off-by: gengliqi <gengliqiii@gmail.com>

### What problem does this PR solve?

Problem Summary:
This PR is split from https://github.com/tikv/tikv/pull/10289.

### What is changed and how it works?

What's Changed:
Always enable entry cache and optimize `term` function a bit.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- No code

Side effects

- No

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/store/peer_storage.rs [hunk 1]
```diff
@@ -1164,64 +1163,50 @@ where
...
     }
 
     pub fn compact_cache_to(&mut self, idx: u64) {
-        if let Some(ref mut cache) = self.cache {
-            cache.compact_to(idx);
-        } else {
-            let rid = self.get_region_id();
+        self.cache.compact_to(idx);
+        let rid = self.get_region_id();
+        if self.engines.raft.has_builtin_entry_cache() {
```
Comment (BusyJay) [inline components/raftstore/src/store/peer_storage.rs:1174]: Can it be removed completely?

File: components/raftstore/src/store/peer_storage.rs [hunk 2]
```diff
@@ -887,8 +876,20 @@ where
         if self.truncated_term() == self.last_term || idx == self.last_index() {
             return Ok(self.last_term);
         }
-        let entries = self.entries(idx, idx + 1, raft::NO_LIMIT)?;
-        Ok(entries[0].get_term())
+        let cache_low = self.cache.first_index().unwrap_or(u64::MAX);
```
Comment (BusyJay) [inline components/raftstore/src/store/peer_storage.rs:879]: How about adding a `entry` method that returns a single entry instead?

**Review by BusyJay — COMMENTED** [review]
Rest LGTM

File: components/raftstore/src/store/peer_storage.rs (hunk 1 above)
Comment (gengliqi) [reply components/raftstore/src/store/peer_storage.rs:1174]: I don't know the future plan for raft engine. /cc @hicqu @tabokie

File: components/raftstore/src/store/peer_storage.rs [hunk 3]
```diff
@@ -887,8 +879,19 @@ where
         if self.truncated_term() == self.last_term || idx == self.last_index() {
             return Ok(self.last_term);
         }
-        let entries = self.entries(idx, idx + 1, raft::NO_LIMIT)?;
-        Ok(entries[0].get_term())
+        if let Some(e) = self.cache.entry(idx) {
+            Ok(e.get_term())
+        } else {
+            let mut entries = vec![];
```
Comment (BusyJay) [inline components/raftstore/src/store/peer_storage.rs:885]: You mean `self.engines.raft.get_entry(self.get_region_id(), idx)`?

**Review by BusyJay — APPROVED** [review]

File: components/raftstore/src/store/peer_storage.rs (hunk 3 above)
Comment (gengliqi) [reply components/raftstore/src/store/peer_storage.rs]: Yep. It's the original code. I change to use `get_entry`.

File: components/raftstore/src/store/peer_storage.rs (hunk 1 above)
Comment (hicqu) [reply components/raftstore/src/store/peer_storage.rs:1174]: Entry cache will be removed from RaftEngine. However for now let's keep the code.

**Review by hicqu — APPROVED** [review]

**Pull request merged**